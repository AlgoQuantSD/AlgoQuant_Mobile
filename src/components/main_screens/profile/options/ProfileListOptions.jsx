import { Ionicons } from "@expo/vector-icons";
import { React, useContext, useEffect, useState } from "react";
import { StyleSheet, Text, View } from "react-native";
import { TouchableOpacity } from "react-native-gesture-handler";
import AlgoquantApiContext from "../../../../general_constants/api/apiContext";
import { WelcomeScreenContext } from "../../../../general_constants/context/WelcomeScreenContext";
import { THEME } from "../../../../general_constants/theme/Theme";
import { handleSignOut } from "../../../../general_helpers/signOut";
import {
  connectToAlpacaModalBuilder,
  deleteAccountModalBuilder,
  disconnectFromAlpacaModalBuilder,
  resetPasswordModalBuilder,
  updateEmailModalBuilder,
  updatePhoneModalBuilder,
} from "../../../general_use/modal/helpers/modalFactory";
import { resetWelcomeScreen } from "../../welcome/helpers/welcomScreenState";

export default function ProfileListOptions(props) {
  // Get access to the algoquant object from parent component to use api
  const algoquantApi = useContext(AlgoquantApiContext);

  // Use to show the welcome screen when the user presses AlgoQuant Tutorial
  const setShowWelcomeScreen = useContext(WelcomeScreenContext);

  // State to keep track if a users account is conntect to alpaca or not
  // Boolean value
  const [alpacaConnection, setAlpacaConnection] = useState(false);
  // Set the information for the corresponding modal
  // (This will be extracted into a separate file containing all functions for setting information for different types of modals)
  function handleListItemPress(key) {
    switch (key) {
      case "TRADE_HISTORY":
        props.navigation.navigate("TradeHistoryScreen");
        break;
      case "CONNECT_ALPACA":
        connectToAlpacaModalBuilder(props);
        break;
      case "DISCONNECT_ALPACA":
        disconnectFromAlpacaModalBuilder(props);
        break;
      case "RESET_PASSWORD":
        resetPasswordModalBuilder(props);
        break;
      case "UPDATE_EMAIL":
        updateEmailModalBuilder(props);
        break;
      case "UPDATE_PHONE":
        updatePhoneModalBuilder(props);
        break;
      case "ALGOQUANT_TUTORIAL":
        resetWelcomeScreen(setShowWelcomeScreen);
        break;
      case "DELETE_ACCOUNT":
        deleteAccountModalBuilder(props);
        break;
      case "SIGN_OUT":
        handleSignOut();
        break;
      default:
    }
  }

  const options = [
    { label: "View trade history", key: "TRADE_HISTORY" },
    { label: "Connect to Alpaca", key: "CONNECT_ALPACA" },
    { label: "Disconnect from Alpaca", key: "DISCONNECT_ALPACA" },
    { label: "Reset password", key: "RESET_PASSWORD" },
    { label: "Update email", key: "UPDATE_EMAIL" },
    { label: "Update phone number", key: "UPDATE_PHONE" },
    { label: "AlgoQuant Tutorial", key: "ALGOQUANT_TUTORIAL" },
    { label: "Delete account", key: "DELETE_ACCOUNT" },
    { label: "Sign out", key: "SIGN_OUT" },
  ];

  // Render the list options based on the users state
  // If the user is connected to Alpaca (true) then DISCONNECTED_ALPACA will show "Disconnect from Alpaca" on the screen
  // If the user is not connected to Alpaca (false) then CONNECT_ALPACA will show "Connect to Alpaca" on the screen
  const filteredOptions = options.filter((option) => {
    if (option.key === "CONNECT_ALPACA") {
      return !alpacaConnection;
    } else if (option.key === "DISCONNECT_ALPACA") {
      return alpacaConnection;
    }
    return true;
  });

  // Get user information using algoquant api, set the alpaca connection based on the information receieved.
  useEffect(() => {
    if (algoquantApi.token) {
      algoquantApi.getUser().then((resp) => {
        setAlpacaConnection(resp.data.alpaca);
      });
    }
  }, [algoquantApi]);

  return (
    <View style={styles.profileListOptionsContainer}>
      {/* Iterate through the options list to render all the options with their labels */}
      {/* If one of the options is pressed we call a handle press function and pass the key to identify which option was pressed */}
      {filteredOptions.map(({ label, key }) => (
        <TouchableOpacity
          style={styles.listItem}
          key={key}
          onPress={() => handleListItemPress(key)}
        >
          <Text style={styles.text}>{label}</Text>
          <Ionicons
            name={THEME.icon.name.arrowRight}
            size={THEME.icon.size.small}
            color={THEME.icon.color.primary}
          />
        </TouchableOpacity>
      ))}
    </View>
  );
}

const styles = StyleSheet.create({
  profileListOptionsContainer: {
    alignItems: "flex-start",
    justifyContent: "flex-start",

    paddingLeft: "10%",
    flex: 0.6,
    width: "100%",
  },
  text: {
    fontSize: THEME.text.fontSize.body,
    color: THEME.text.color.primary,
    paddingRight: "1%",
  },
  listItem: {
    flexDirection: "row",
    alignItems: "center",
    paddingBottom: "9%",
  },
});
