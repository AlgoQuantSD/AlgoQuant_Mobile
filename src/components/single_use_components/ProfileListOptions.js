import React from "react";
import { View, Text, StyleSheet } from "react-native";
import { THEME } from "../../constants/Theme";
import { Ionicons } from "@expo/vector-icons";
import { TouchableOpacity } from "react-native-gesture-handler";
import {
  connectToAlpacaModalBuilder,
  deleteAccountModalBuilder,
  disconnectFromAlpacaModalBuilder,
} from "../../helpers/modalFactory";
import { handleSignOut } from "../../helpers/signOut";
import { MOCK_USER } from "../../constants/MockUser";

export default function ProfileListOptions(props, { navigation }) {
  // Set the information for the corresponding modal
  // (This will be extracted into a separate file containing all functions for setting information for different types of modals)
  function handleListItemPress(key) {
    switch (key) {
      case "CONNECT_ALPACA":
        connectToAlpacaModalBuilder(props);
        break;
      case "DISCONNECT_ALPACA":
        disconnectFromAlpacaModalBuilder(props);
        break;
      case "DELETE_ACCOUNT":
        deleteAccountModalBuilder(props);
        break;
      case "SIGN_OUT":
        handleSignOut();
        break;
      default:
    }
  }

  const options = [
    { label: "View trade history", key: "TRADE_HISTORY" },
    { label: "Connect to Alpaca", key: "CONNECT_ALPACA" },
    { label: "Disconnect from Alpaca", key: "DISCONNECT_ALPACA" },
    { label: "Reset password", key: "RESET_PASSWORD" },
    { label: "Update email", key: "UPDATE_EMAIL" },
    { label: "Delete account", key: "DELETE_ACCOUNT" },
    { label: "Sign out", key: "SIGN_OUT" },
  ];

  // Render the list options based on the users state
  // We don't want to render connect to alpaca if the user is already connected
  const filteredOptions = options.filter((option) => {
    if (option.key === "CONNECT_ALPACA") {
      return !MOCK_USER.alpaca.isConnected;
    } else if (option.key === "DISCONNECT_ALPACA") {
      return MOCK_USER.alpaca.isConnected;
    }
    return true;
  });

  return (
    <View style={styles.profileListOptionsContainer}>
      {/* Iterate through the options list to render all the options with their labels */}
      {/* If one of the options is pressed we call a handle press function and pass the key to identify which option was pressed */}
      {filteredOptions.map(({ label, key }) => (
        <TouchableOpacity
          style={styles.listItem}
          key={key}
          onPress={() => handleListItemPress(key)}
        >
          <Text style={styles.text}>{label}</Text>
          <Ionicons
            name="arrow-forward"
            size={THEME.text.fontSizeBody}
            color={THEME.colors.foreground}
          />
        </TouchableOpacity>
      ))}
    </View>
  );
}

const styles = StyleSheet.create({
  profileListOptionsContainer: {
    alignItems: "flex-start",
    justifyContent: "flex-start",
    paddingTop: "5%",
    paddingLeft: "10%",
    flex: 0.6,
    width: "100%",
  },
  text: {
    fontSize: THEME.text.fontSizeBody,
    color: THEME.text.color,
    paddingRight: "1%",
  },
  listItem: {
    flexDirection: "row",
    alignItems: "center",
    paddingBottom: "10%",
  },
});
