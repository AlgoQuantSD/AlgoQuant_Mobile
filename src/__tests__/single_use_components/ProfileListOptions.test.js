import React from "react";
import { render, screen } from "@testing-library/react-native";
import ProfileListOptions from "../../components/single_use_components/ProfileListOptions";

describe("ProfileListOptions", () => {
  it("Should render", () => {
    const tree = render(<ProfileListOptions />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("renders the correct list options", () => {
    render(<ProfileListOptions />);
    const listOptions = [
      "View trade history",
      "Connect to Alpaca",
      "Reset password",
      "Update email",
      "Delete account",
      "Sign out",
    ];
    listOptions.forEach((option) => {
      screen.getByText(option);
    });
  });
});
