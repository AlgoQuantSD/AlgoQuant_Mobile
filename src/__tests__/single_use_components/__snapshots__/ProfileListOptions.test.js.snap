// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileListOptions Should render 1`] = `
<View
  style={
    Object {
      "alignItems": "flex-start",
      "flex": 0.6,
      "justifyContent": "flex-start",
      "paddingLeft": "10%",
      "paddingTop": "5%",
      "width": "100%",
    }
  }
>
  <RNGestureHandlerButton
    collapsable={false}
    delayLongPress={600}
    enabled={true}
    exclusive={true}
    handlerTag={1}
    handlerType="NativeViewGestureHandler"
    onGestureEvent={[Function]}
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    onHandlerStateChange={[Function]}
    rippleColor={0}
    touchSoundDisabled={false}
  >
    <View
      accessible={true}
      collapsable={false}
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "opacity": 1,
          "paddingBottom": "10%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#FFFFFF",
            "fontSize": 16,
            "paddingRight": "1%",
          }
        }
      >
        View trade history
      </Text>
      <Text />
    </View>
  </RNGestureHandlerButton>
  <RNGestureHandlerButton
    collapsable={false}
    delayLongPress={600}
    enabled={true}
    exclusive={true}
    handlerTag={2}
    handlerType="NativeViewGestureHandler"
    onGestureEvent={[Function]}
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    onHandlerStateChange={[Function]}
    rippleColor={0}
    touchSoundDisabled={false}
  >
    <View
      accessible={true}
      collapsable={false}
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "opacity": 1,
          "paddingBottom": "10%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#FFFFFF",
            "fontSize": 16,
            "paddingRight": "1%",
          }
        }
      >
        Connect to Alpaca
      </Text>
      <Text />
    </View>
  </RNGestureHandlerButton>
  <RNGestureHandlerButton
    collapsable={false}
    delayLongPress={600}
    enabled={true}
    exclusive={true}
    handlerTag={3}
    handlerType="NativeViewGestureHandler"
    onGestureEvent={[Function]}
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    onHandlerStateChange={[Function]}
    rippleColor={0}
    touchSoundDisabled={false}
  >
    <View
      accessible={true}
      collapsable={false}
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "opacity": 1,
          "paddingBottom": "10%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#FFFFFF",
            "fontSize": 16,
            "paddingRight": "1%",
          }
        }
      >
        Reset password
      </Text>
      <Text />
    </View>
  </RNGestureHandlerButton>
  <RNGestureHandlerButton
    collapsable={false}
    delayLongPress={600}
    enabled={true}
    exclusive={true}
    handlerTag={4}
    handlerType="NativeViewGestureHandler"
    onGestureEvent={[Function]}
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    onHandlerStateChange={[Function]}
    rippleColor={0}
    touchSoundDisabled={false}
  >
    <View
      accessible={true}
      collapsable={false}
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "opacity": 1,
          "paddingBottom": "10%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#FFFFFF",
            "fontSize": 16,
            "paddingRight": "1%",
          }
        }
      >
        Update email
      </Text>
      <Text />
    </View>
  </RNGestureHandlerButton>
  <RNGestureHandlerButton
    collapsable={false}
    delayLongPress={600}
    enabled={true}
    exclusive={true}
    handlerTag={5}
    handlerType="NativeViewGestureHandler"
    onGestureEvent={[Function]}
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    onHandlerStateChange={[Function]}
    rippleColor={0}
    touchSoundDisabled={false}
  >
    <View
      accessible={true}
      collapsable={false}
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "opacity": 1,
          "paddingBottom": "10%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#FFFFFF",
            "fontSize": 16,
            "paddingRight": "1%",
          }
        }
      >
        Delete account
      </Text>
      <Text />
    </View>
  </RNGestureHandlerButton>
  <RNGestureHandlerButton
    collapsable={false}
    delayLongPress={600}
    enabled={true}
    exclusive={true}
    handlerTag={6}
    handlerType="NativeViewGestureHandler"
    onGestureEvent={[Function]}
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    onHandlerStateChange={[Function]}
    rippleColor={0}
    touchSoundDisabled={false}
  >
    <View
      accessible={true}
      collapsable={false}
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "opacity": 1,
          "paddingBottom": "10%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#FFFFFF",
            "fontSize": 16,
            "paddingRight": "1%",
          }
        }
      >
        Sign out
      </Text>
      <Text />
    </View>
  </RNGestureHandlerButton>
</View>
`;
